### ISSUE TITLE
[Bug]: CPU MoE kernel produces NaN logits under torch.compile for Qwen3_5MoeForConditionalGeneration — reproduces with AND without GPTQ-Int4 quantization, fixed by --enforce-eager

### ISSUE THREAD

**Issue by ixcans** [vllm-project/vllm#55560]
> **Update (2026-09-06, same day):** the original title/body below claimed this
> was specific to the GPTQ-Int4 quantized kernel path. That was **wrong** —
> further testing reproduced the identical bug on the **unquantized bf16**
> release of the same checkpoint (no `quantize_config.json`, no GPTQ at all).
> `--enforce-eager` fixes it in both cases. So this is a general CPU MoE +
> torch.compile/inductor issue for this model architecture
> (`Qwen3_5MoeForConditionalGeneration`), independent of quantization method
> — not a GPTQ-Int4-specific dequantization/kernel bug as originally
> reported. Title and body below have been corrected in place (not left as a
> stale/wrong report) to reflect this. The GPTQ-Int4 repro is kept since it's
> still a valid, minimal reproduction; the bf16 repro is added alongside it.

### Your current environment

- vLLM CPU-backend image: `vllm/vllm-openai-cpu@sha256:ea8b61f8ccab650e1c369105373863ac12a581b53f289b344cea72a09238e131` (v0.28.0 — the latest published x86_64 CPU image at the time of testing)
- Hardware: bare-metal Intel Xeon (AMX-capable: `amx_bf16`, `amx_int8`, `amx_tile` all present), CPU-only inference (no GPU)
- Backend: `VLLM_CPU_KVCACHE_SPACE=16`, `VLLM_CPU_OMP_THREADS_BIND=auto`, `dtype=bfloat16`, `max_model_len=8192`, `enable_chunked_prefill=True`, TP=1

### Model Input Dumps

Reproduces on **both** of these (same base model, same architecture, different quantization):

- [`llmfan46/Qwen3.6-35B-A3B-uncensored-heretic-Native-MTP-Preserved-GPTQ-Int4`](https://huggingface.co/llmfan46/Qwen3.6-35B-A3B-uncensored-heretic-Native-MTP-Preserved-GPTQ-Int4) (revision `fb685a8409e4290f8a15dad0a691e6a9e3d42c3f`) — GPTQ 4-bit quantization auto-detected from `quantize_config.json`.
- [`llmfan46/Qwen3.6-35B-A3B-uncensored-heretic-Native-MTP-Preserved`](https://huggingface.co/llmfan46/Qwen3.6-35B-A3B-uncensored-heretic-Native-MTP-Preserved) (revision `9599ac17d26daf33daf0fdd8f6c897ff4c6dc89a`) — **unquantized bf16**, no `quantize_config.json` at all.

Both: architecture `Qwen3_5MoeForConditionalGeneration` (MoE, hybrid linear/full attention).

### 🐛 Describe the bug

With the default torch.compile/inductor compilation path enabled
(`enforce_eager=False`), the CPU MoE kernel path for this architecture
produces **NaN logits** for every request, regardless of prompt or endpoint,
and **regardless of whether the checkpoint is GPTQ-Int4 quantized or
unquantized bf16**. This manifests as:

- `/v1/chat/completions` on a simple prompt ("What is 2 + 2? Reply with only the number.") returns a degenerate, repeated `"!"` completion (e.g. `"!!!!!!!!!!!!!!!!"`) and never reaches EOS — `finish_reason: "length"` even at very low max_tokens.
- A raw `/v1/completions` call on `"The capital of France is"` (bypassing the chat template entirely) produces the same degenerate `"!"` output.
- Requesting `logprobs=5` on either endpoint surfaces the actual defect: `"Out of range float values are not JSON compliant: nan"` — the sampled logits are NaN, not merely a bad sample.

The model **loads successfully** and passes health/readiness checks (`/health` returns 200) in both cases, which makes this a silent-failure mode worth flagging on its own — a healthy pod/process gives zero signal that generation is broken.

**Workaround: `--enforce-eager` completely fixes it, on both checkpoints.** With the identical checkpoint, config, and request, eager mode produces correct, coherent generations with real (non-NaN) logprobs and a proper EOS (`finish_reason: "stop"`). This isolates the defect to a torch.compile/inductor miscompilation of the CPU MoE code path for this architecture — not a GPTQ dequantization correctness bug (ruled out by the bf16 repro having no quantization at all), and not the CPU-MoE-GPTQ support gap tracked in #37979 (that issue is about whether MoE+GPTQ support exists on CPU at all; here it exists and works on both quantized and unquantized checkpoints — only the compiled path is broken).

Isolated MoE specifically (rather than this model's hybrid linear-attention architecture, or GPTQ quantization) as the trigger by comparing against a sibling model that shares the same `Qwen3_5*ForConditionalGeneration` hybrid-attention family but is dense (no MoE) and uses INT8 compressed-tensors — that sibling works correctly with torch.compile enabled (default settings, no `--enforce-eager` needed). Combined with the GPTQ-vs-bf16 result above, this narrows the trigger to "MoE, on CPU, under torch.compile/inductor, for this architecture" — independent of quantization.

### Steps to reproduce

**GPTQ-Int4 checkpoint:**
```
vllm serve llmfan46/Qwen3.6-35B-A3B-uncensored-heretic-Native-MTP-Preserved-GPTQ-Int4 \
  --revision fb685a8409e4290f8a15dad0a691e6a9e3d42c3f \
  --dtype bfloat16 --max-model-len 8192 --enable-chunked-prefill \
  --served-model-name qwen3-6-35b-a3b-gptq-int4
```

**Unquantized bf16 checkpoint (same bug, same fix):**
```
vllm serve llmfan46/Qwen3.6-35B-A3B-uncensored-heretic-Native-MTP-Preserved \
  --revision 9599ac17d26daf33daf0fdd8f6c897ff4c6dc89a \
  --dtype bfloat16 --max-model-len 8192 --enable-chunked-prefill \
  --served-model-name qwen3-6-35b-a3b-bf16
```

Then, for either:

1. Send a completion request with `logprobs: 5`:
   ```
   curl -s http://localhost:8000/v1/completions -H 'Content-Type: application/json' -d '{
     "model": "<served-model-name-above>",
     "prompt": "The capital of France is",
     "max_tokens": 16,
     "logprobs": 5
   }'
   ```
   Observe: degenerate `"!"` output, and/or a JSON-serialization error surfacing NaN logprobs.
2. Re-run with `--enforce-eager` added to the serve command. Observe: coherent completion (`" Paris.\nThe capital of France is Paris...."`), real logprobs, no NaN.

### Expected behavior

Generation should be correct with torch.compile enabled (the default), matching eager-mode output, for this CPU MoE checkpoint — whether GPTQ-Int4 quantized or unquantized.

### Additional context

Not yet tested: whether this reproduces on other MoE architectures on the CPU backend, or is specific to `Qwen3_5MoeForConditionalGeneration`. Filing as-is since both repros above are clean and the workaround is known — happy to help narrow it further if useful.

Related: #37979 (tracks CPU MoE+GPTQ support landing at all — this bug is in the already-working support path, on both quantized and unquantized checkpoints, once torch.compile is disabled).

**Comment by ixcans** [comment]
no longer reproduces on the v0.29.0
